tests/testsuite/open_namespaces.rs

```diff
@@ -0,0 +1,358 @@
+use cargo_test_support::project;
+use cargo_test_support::registry::RegistryBuilder;
+
+#[cargo_test]
+fn within_namespace_requires_feature() {
+    let p = project()
+        .file(
+            "Cargo.toml",
+            r#"
+                [package]
+                name = "foo::bar"
+                version = "0.0.1"
+                edition = "2015"
+            "#,
+        )
+        .file("src/lib.rs", "")
+        .build();
+
+    p.cargo("read-manifest")
+        .masquerade_as_nightly_cargo(&["open-namespaces"])
+        .with_status(101)
+        .with_stderr(
+            r#"error: failed to parse manifest at `[CWD]/Cargo.toml`
+
+Caused by:
+  feature `open-namespaces` is required
+
+  The package requires the Cargo feature called `open-namespaces`, but that feature is not stabilized in this version of Cargo ([..]).
+  Consider adding `cargo-features = ["open-namespaces"]` to the top of Cargo.toml (above the [package] table) to tell Cargo you are opting in to use this unstable feature.
+  See https://doc.rust-lang.org/nightly/cargo/reference/unstable.html#open-namespaces for more information about the status of this feature.
+"#,
+        )
+        .run()
+}
+
+#[cargo_test]
+fn implicit_lib_within_namespace() {
+    let p = project()
+        .file(
+            "Cargo.toml",
+            r#"
+                cargo-features = ["open-namespaces"]
+
+                [package]
+                name = "foo::bar"
+                version = "0.0.1"
+                edition = "2015"
+            "#,
+        )
+        .file("src/lib.rs", "")
+        .build();
+
+    p.cargo("read-manifest")
+        .masquerade_as_nightly_cargo(&["open-namespaces"])
+        .with_json(
+            r#"{
+  "authors": [],
+  "categories": [],
+  "default_run": null,
+  "dependencies": [],
+  "description": null,
+  "documentation": null,
+  "edition": "2015",
+  "features": {},
+  "homepage": null,
+  "id": "path+file://[..]#foo::bar@0.0.1",
+  "keywords": [],
+  "license": null,
+  "license_file": null,
+  "links": null,
+  "manifest_path": "[CWD]/Cargo.toml",
+  "metadata": null,
+  "name": "foo::bar",
+  "publish": null,
+  "readme": null,
+  "repository": null,
+  "rust_version": null,
+  "source": null,
+  "targets": [
+    {
+      "crate_types": [
+        "lib"
+      ],
+      "doc": true,
+      "doctest": true,
+      "edition": "2015",
+      "kind": [
+        "lib"
+      ],
+      "name": "foo::bar",
```

Comment:
My understanding from the RFC was that the library name was meant to just be `bar`, that would avoid the rustc issue you noted

> The autogenerated `lib.name` key for such a crate will just be bar, the leaf crate name, and the expectation is that to use such crates one must use `--extern foo::bar=bar.rlib` syntax. There may be some better things possible here, perhaps `foo_bar` can be used here.
> —<https://rust-lang.github.io/rfcs/3243-packages-as-optional-namespaces.html#reference-level-explanation>

Reply:
(I'd assume this also applies to the autogenerated `bin.name` for `src/main.rs`)